tests/unit/test_llm_router.py

```diff
@@ -0,0 +1,514 @@
+            },
+        )
+    ]
+    return LLMConfig(
+        model='gpt-4-vision-enhanced',
+        router_config=RouterConfig(
+            model_list=model_list,
+            routing_strategy='simple-shuffle',
+            cooldown_time=1.0,
+            num_retries=8,
+            retry_after=15,
+            allowed_fails=2,
+            cache_responses=False,
+        ),
+    )
+
+
+##############################################################################
+##############################################################################
+
+
+@patch('openhands.llm.llm.time.sleep')
+@patch('openhands.llm.llm.litellm.Router')
+def test_router_completion_rate_limit_wait_time(
+    mock_router, mock_sleep, llm_router_config
+):
+    # Set up the mock router's completion method to simulate rate limiting
+    mock_router_instance = mock_router.return_value
+    mock_completion = mock_router_instance.completion
+
+    # Simulate rate limiting on the first attempt and a successful completion on the second attempt
+    mock_completion.side_effect = [
+        RateLimitError(
+            message='Rate limit exceeded',
+            llm_provider='test_provider',
+            model='test_model',
+            response=None,
+            litellm_debug_info=None,
+            max_retries=2,
+            num_retries=1,
+        ),
+        {
+            'choices': [{'message': {'content': 'Retry successful'}}]
+        },  # Successful response on the second call
+    ]
+    # Set num_retries to 1 to prevent retry logic from interfering
+    llm_router_config.num_retries = 1
+
+    llm = LLM(config=llm_router_config)
+
+    # Expect a RateLimitError to be raised since num_retries is 1
+    with pytest.raises(RateLimitError):
+        llm.completion(
+            messages=[{'role': 'user', 'content': 'Hello!'}],
+            stream=False,
+        )
+
+    # Assertions
+    assert mock_completion.call_count == 1
+    mock_sleep.assert_not_called()
+
+
+@pytest.mark.parametrize(
+    'exception_type, extra_args',
+    [
+        (APIConnectionError, {}),
+        (InternalServerError, {}),
+        (ServiceUnavailableError, {}),
+        (RateLimitError, {'response': None}),  # response is required for RateLimitError
+    ],
+)
+@patch('openhands.llm.llm.litellm.Router')
+def test_router_completion_retries(
+    mock_router, llm_router_config, exception_type, extra_args
+):
+    """Since the Router class handles retries internally, when it exhausts all retries,
+    it raises an exception. In this test, we're mocking the completion method to raise
+    an exception on the first call and return a successful response on the second.
+    However, the Router class is designed to retry only when the exception is
+    RateLimitError. For other exceptions, it will not retry, unless there is a custom
+    retry policy defined in the RouterConfig.
```

Comment:
🤔 I saw AuthErrors and a message from litellm saying it retried 2 times and failed again the 3rd. I didn't have a retry policy with auth errors. It looked as if it retries everything if there is no custom policy.